### DISCUSSION TITLE
Example for a 2D Median Filter

### DISCUSSION THREAD

**Discussion by RoyiAvital** [stev47/StaticKernels.jl#12] [General]
I find `StaticKernels.jl` to be a jewel for low level image processing in Julia.  
It is odd, as Julia is very good choice for image processing yet the landscape of packages is rather small.
Unless you go the `JuliaImages.jl` path, which I don't like due to the choice Arrays of Struct as the underlying type and large dependency.
I prefer working with `Matrix{T}` and `Array{T, 3}` for images.

This is how I implemented Median Filter with `StaticKernels.jl`:

```
mI = rand(10, 10); #<! Image
mK = Kernel{(-2:2, -2:2)}(@inline w -> median(Tuple(w))); #<! The kernel (Window function)
mF = map(mK, mI); #<! Filtered image
``` 

The `Tuple(w)` is required as in #8 (Iterating the window).

There are optimized algorithms for 2D Median, yet I couldn't find implementation for simple arrays in Julia.

**Comment by stev47** [comment]
I think you meant `mF = map(mK, MI)` (kernel first) in your last line.

For some kernels like median there are more clever algorithms which scale better with window size such as the [Herk-Gil-Werman algorithm](https://ieeexplore.ieee.org/abstract/document/211471). Keep in mind that `StaticKernels` will not be able to do that.

The [MR implementing median for `StaticArrays`](https://github.com/JuliaArrays/StaticArrays.jl/pull/973) is still open, but you can hand-roll your own median e.g. like this
```julia
@inline function mymedian(w::StaticKernels.Window)
  v = SVector(Tuple(w))
  v = @inline sort(v; alg = StaticArrays.BitonicSort)
  isodd(length(v)) ? v[end ÷ 2 + 1] : ((v[end ÷ 2] + v[end ÷ 2 + 1]) / 2)
end
k = Kernel{(-2:2,-2:2)}(mymedian)
```
which should be slightly faster than what you did.

**Reply by RoyiAvital** [reply]
Yes, I am aware of the more efficient way for rolling medians (I wrote that above).  
But in the current eco system it means one must implement by himself.

Your package gives me the ability to do things quick.  
So it is better to have something than potentially better :-).

So this is just a post for appreciation and others, in case of need, will be able to use.
